### PR TITLE
Support big- and little-endian lane order with bitcast

### PR DESCRIPTION
**Pull request by uweigand**
Add a MemFlags operand to the bitcast instruction, where only the `big` and `little` flags are accepted.  These define the lane order to be used when casting between types of different lane counts.

Update all users to pass an appropriate MemFlags argument.

Implement lane swaps where necessary in the s390x back-end.

This is the final part necessary to fix
https://github.com/bytecodealliance/wasmtime/issues/4566.

CC @cfallin 

This is still a RFC because of the following questions:
- Should we *enforce* that a byte order flag must be explicitly specified if the types differ in lane count?  (This would be easily doable in the verifier.)
- There was some discussion in the issue whether or not the `MemFlags` API is the right choice.  It seems that way to me, since it makes the semantics easy to specific, and is simple to implement based on the already existing `MemFlags` infrastructure.  If you prefer some other implementation, please let me know - I'd be happy to change this.

### CODE REVIEW

File: cranelift/codegen/meta/src/shared/instructions.rs [hunk 1]
```diff
@@ -3113,11 +3114,15 @@ pub(crate) fn define(
 
         The input and output types must be storable to memory and of the same
         size. A bitcast is equivalent to storing one type and loading the other
-        type from the same address.
+        type from the same address, using the specified MemFlags.
```
Comment (cfallin) [inline cranelift/codegen/meta/src/shared/instructions.rs:3117]: s/using/both using/, just for clarity (other possibilities here include storing with one `MemFlags` and loading with another, which isn't what we're emulating but someone might imagine this  instead)

**Review by cfallin — COMMENTED** [review]
Thanks -- this is surprisingly simple; I'm happy how cleanly it came out. The simple definition/derivation of bitcast's semantics via store+load combines really nicely with our endianness support generally.

I think using a restricted subset of `MemFlags` is fine for that reason. (Instruction format too; at first `LoadNoOffset` seemed a bit odd but none of the instruction predicates come directly from the format (e.g. `can_load`) and so I think it's better than introducing another enum variant in the end.)

I *do* think we should require one of `big` or `little` when lane-count differs, as you suggest; otherwise this is a new way that endian-specific behavior is visible in CLIF, which I'd prefer to avoid. Happy to r+ once that is added!

**Comment by uweigand** [comment]
> I _do_ think we should require one of `big` or `little` when lane-count differs, as you suggest; otherwise this is a new way that endian-specific behavior is visible in CLIF, which I'd prefer to avoid. Happy to r+ once that is added!

Ok, implemented.  There were four filetests where the error triggered, which I fixed by adding a byte order specifier.  Thanks!

**Review by cfallin — APPROVED** [review]
LGTM with changes, thanks!

**Pull request merged**